loky/backend/_thread_pool_limiters.py

```diff
@@ -0,0 +1,443 @@
+    """Limit the number of threads available for threadpools in supported C-lib
+
+    Set the maximal number of thread that can be used in thread pools used in
+    the supported C-libraries to `max_threads_per_process`. This function works
+    for libraries that are already loaded in the interpreter and can be changed
+    dynamically.
+
+    The `limits` parameter can be either an interger or a dict to specify the
+    maximal number of thread that can be used in thread pools. If it is an
+    integer, sets the maximum number of thread to `limits` for each C-lib
+    selected by `subset`. If it is a dictionary
+    `{supported_libraries: max_threads}`, this function sets a custom maximum
+    number of thread for each C-lib. If None, does not do anything.
+
+    The `subset` parameter select a subset of C-libs to limit. Used only if
+    `limits` is an int. If it is "all" or None, this function will limit all
+    supported C-libs. If it is "blas", it will limit only BLAS supported
+    C-libs and if it is "openmp", only only OpenMP supported C-libs will be
+    limited. Note that the latter can affect the number of threads used by the
+    BLAS C-libs if they rely on OpenMP.
+
+    If `reload_clib` is `True`, first loop through the loaded libraries to
+    ensure that this function is called on all available libraries.
+
+    Return a dict dynamic_threadpool_size containing pairs `('clib': boolean)`
+    which are True if `clib` have been found and can be used to scale the
+    maximal number of threads dynamically.
+    """
+    wrapper = _get_wrapper(reload_clib)
+    return wrapper.set_thread_limits(limits, subset)
+
+
+def get_thread_limits(reload_clib=False):
+    """Return maximal thread number for threadpools in supported C-lib
+
+    Return a dictionary containing the maximal number of threads that can be
+    used in supported libraries or None when the library is not available. The
+    key of the dictionary are {}.
+
+    If `reload_clib` is `True`, first loop through the loaded libraries to
+    ensure that this function is called on all available libraries.
+    """.format(list(_CLibsWrapper.SUPPORTED_CLIBS.keys()))
+    wrapper = _get_wrapper(reload_clib)
+    return wrapper.get_thread_limits()
+
+
+class thread_pool_limits:
+    """Change the default number of threads used in thread pools.
+
+    This class can be used either as a function (the construction of this
+    object limits the number of threads) or as a context manager, in a `with`
+    block.
+
+    Parameters
+    ----------
+    limits : int or dict, (default=None)
+        Maximum number of thread that can be used in thread pools
+
+        If int, sets the maximum number of thread to `limits` for each C-lib
+        selected by `subset`.
+
+        If dict(supported_libraries: max_threads), sets a custom maximum number
+        of thread for each C-lib.
+
+        If None, does not do anything.
+
+    subset : string or None, optional (default="all")
```

Comment:
This could be renamed to "api".

Reply:
done in `apis` as it is a subset. My main concern is that we are not consistent with the internal `api` (`omp`, `openblas` and `mkl`) and these one (`all`, `blas`, `openmp`).
Do you have a better name for this?

Reply:
I agree that api is well suited for the internal clibs apis, e.g. openmp_set_num_threads etc...
As api is already used elsewhere, I think it's confusing to mix api and apis keywords.

Here we want to select a subset of the clibs so subset seemed fine. I also find `libs` explicit

Reply:
To me "API" is better reserved for public APIs like BLAS and OpenMP.

- OpenBLAS and MKL are implementations of BLAS.
- gomp, vomp, iomp, libomp are implementations of the OpenMP API.

We should dispatch to the right g/set_num_threads  internal APIs based on the implementations instead (when they are implementation specific as is the case for BLAS implementations).

The resulting code would be more readable.

Reply:
Technically speaking OpenMP is an API and BLAS is a specification :)